### ISSUE TITLE
Feat: Explicit list for dependencies version for all extras (dev,db-related) and build packages

### ISSUE THREAD

**Issue by inkhey** [tracim/tracim#3191]
## Feature description

As we've seen in https://github.com/tracim/tracim/issues/3025, development dependencies versions can also break the CI and others things.

## Chosen solution

After some search, we see that it's not possible to use pip-compile to support strictly these case as pip-compile doesn't cope properly with both build packages (pip, setuptools, wheel) and extras dependencies. (see comments below)
So, for a first step, we should provide static list of first level dependencies for extras and build packages, in order to be able to provide the most static environnement possible if needed.

We should use these static versions in docker, backend setup script and travis ci.

Futures list of requirements files:

-    requirements.txt (handled automatically by pip-tools from setup.py version) : base requirements of production tracim
-    ~~requirements-static-tests.txt : for static tests (travis), already exist~~
-    requirements-dev.txt : only dev specific dependencies (dependencies installed by ".[dev]")
-    requirements-test.txt
-   requirements-devtool.txt
-    requirements-db-postgres.txt : psycopg2 version
-    requirements-db-mysql.txt: PyMySQL[rsa] version
-    requirements-build.txt: pip, setuptools, wheel version

~~Each should contain include of requirements.txt (except  requirements.txt itself of course).~~

original issue for build package : https://github.com/tracim/tracim/issues/3307

## Implemented solution:

- add a requirements file for build dependencies
- add a requirements files for devtool
- add a requirement files for test dependencies
- add 2 requirements files for both mysql and postgresql specififc requirements
- add requirements-dev.txt with is a shortcut for requirements.txt, devtool and test dependencies.
- adapt docker, travis ci and setup_backend script context to use theses requirements file with strict versions of packages

**Comment by inkhey** [comment]
the logical way to do this will be to use pip-compile to compile result of:
`pip install -e ".[dev]"`
but pip-tools (which contain pip-compile) doesn't support `extra_require` feature we do use.
see https://github.com/jazzband/pip-tools/issues/625.

I did some tests with hack proposed with "-e" but didn't success:

dev.in :
`
-e .
`

```
$ pip-compile dev.in
Traceback (most recent call last):
  File "/home/user/tracim/backend/env/bin/pip-compile", line 8, in <module>
    sys.exit(cli())
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/piptools/scripts/compile.py", line 392, in cli
    results = resolver.resolve(max_rounds=max_rounds)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/piptools/resolver.py", line 169, in resolve
    has_changed, best_matches = self._resolve_one_round()
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/piptools/resolver.py", line 255, in _resolve_one_round
    constraints = sorted(self.constraints, key=key_from_ireq)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/piptools/resolver.py", line 128, in constraints
    self._group_constraints(chain(self.our_constraints, self.their_constraints))
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/piptools/resolver.py", line 231, in _group_constraints
    self.repository.get_dependencies(ireq)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/piptools/repositories/pypi.py", line 228, in get_dependencies
    download_dir, ireq, wheel_cache
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/piptools/repositories/pypi.py", line 180, in resolve_reqs
    results = resolver._resolve_one(reqset, ireq)
  File "/home/user/tracim/backend/env/lib/python3.6/site-packages/pip/_internal/resolution/legacy/resolver.py", line 425, in _resolve_one
    assert req_to_install.user_supplied
```

**Comment by inkhey** [comment]
as pip-compile doesn't work here, i suggest to use a simple stricted version list, like `requirements-static-tests.txt`. this is less bullet proof as we don't deal with dependencies version, and do same for postgres, mysql, build dependencies.

Another solution may be:
- remove or keep extras require (but duplicated information)
- create .in file with only dependencies for tests, one another for static-test, another for postgres, another for mysql
and do process .txt from these. The difficulty is with already existing dependencies that will not follow the current requirements.txt version (we do need to check both file to have similar version)
- create a build packages (setuptools, pip, wheel, etc..) requirement .txt file as pip-compile does mix well with theses

**Comment by grignards** [comment]
@inkhey Where would you put the static versions? In separate requirements.txt files? If yes I would include the "base" requirements file in the separate ones.

**Comment by inkhey** [comment]
> @inkhey Where would you put the static versions? In separate requirements.txt files? If yes I would include the "base" requirements file in the separate ones.

Yes, idea is that we will have:
- requirements.txt (handled automatically by pip-tools from setup.py version) : base requirements of production tracim
- requirements-static-tests.txt : for static tests (travis), already exist
- requirements-dev.txt : only dev specific dependencies (dependencies installed by ".[dev]")
- requirements-db-postgres.txt : psycopg2 version
- requirements-db-mysql.txt: PyMySQL[rsa] version
- requirements-build.txt: pip, setuptools, wheel version

I'm not sure what you mean by include, do you mean adding:
`
-r requirements.txt
`
to all requirements file (except requirements.txt itself) ?

**Comment by grignards** [comment]
> I'm not sure what you mean by include, do you mean adding:
> `-r requirements.txt`
> to all requirements file (except requirements.txt itself) ?

Yes this is what I meant.

**Comment by inkhey** [comment]
> > I'm not sure what you mean by include, do you mean adding:
> > `-r requirements.txt`
> > to all requirements file (except requirements.txt itself) ?
> 
> Yes this is what I meant.

That's ok.

**Comment by inkhey** [comment]
look like we got some trouble with dev dependencies: https://github.com/tracim/tracim/pull/3476/checks?check_run_id=1050305685
- current dev environnement use python > 3.5 because : 
  - use more modern distribution
  - black work only with 3.6+

This lead to 2 problem with strict version in requirements.txt for dev in travis:
- some installed version of development package used by dev environnement doesn't exist in python 3.5 like mock, only solution seems downgrade these, see https://github.com/tracim/tracim/pull/3476/checks?check_run_id=1050305685 error
- black need 3.6+, should we dropped it from list and use latest version instead (which is unsure) ? Should we do something else ?

**Comment by grignards** [comment]
>     * some installed version of development package used by dev environnement doesn't exist in python 3.5 like mock, only solution seems downgrade these, see https://github.com/tracim/tracim/pull/3476/checks?check_run_id=1050305685 error

As long as the last version compatible with python 3.5 works for us this could be a solution.

>     * black need 3.6+, should we dropped it from list and use latest version instead (which is unsure) ? Should we do something else ?

black is already part of requirements-static-tests, must it also be in requirements-dev?

**Comment by lebouquetin** [comment]
@inkhey black is a dev tool, not a real dependency, I think.
What is the status of python 3.5 support in tracim?

**Comment by inkhey** [comment]
> @inkhey black is a dev tool, not a real dependency, I think.

> > ```
> > * black need 3.6+, should we dropped it from list and use latest version instead (which is unsure) ? Should we do something else ?
> > ```
> 
> black is already part of requirements-static-tests, must it also be in requirements-dev?

Maybe should i change the may requirements.txt files are splitted : 
 `requirements-static-tests` is currently CI specific, what does exist in backend is testing and devtool, dev mean testing + devtool, maybe should i adapt static-tests to be same as devtool and create a test requirements.

> What is the status of python 3.5 support in tracim?

Python 3.5 is used by stretch based docker, so we do support it. The only blocking point here is for dev requirements which should not be provided in docker/ production environnement.

**Comment by raphj** [comment]
- We've been bitten by an uncontrolled, automatic upgrade of a linter that broke the CI unexpectedly (flake8, https://github.com/tracim/tracim/issues/3025) and prompted an emergency response from us. I don't think reverting to use the latest version of black is a solution.
- using different versions of libraries under different versions of Python, while probably necessary sounds like a more complex system to deal with, with potentially bugs appearing in specific versions of Python. If we can avoid having different versions of dependencies.

Note: the final version of Python 3.5 will be released this Friday (September, 4). Support is ending for this version.

**Comment by grignards** [comment]
> Maybe should i change the mayway requirements.txt files are splitted :
> requirements-static-tests is currently CI specific, what does exist in backend is testing and devtool, dev mean testing + devtool, maybe should i adapt static-tests to be same as devtool and create a requirements-tests.txt.

Yes I think this is a good idea, then dev will only be a shortcut (including) static-tests,tests and build. We'll get 7 requirements.txt files if I'm not wrong…

**Comment by inkhey** [comment]
> > Maybe should i change the mayway requirements.txt files are splitted :
> > requirements-static-tests is currently CI specific, what does exist in backend is testing and devtool, dev mean testing + devtool, maybe should i adapt static-tests to be same as devtool and create a requirements-tests.txt.
> 
> Yes I think this is a good idea, then dev will only be a shortcut (including) static-tests,tests and build. We'll get 7 requirements.txt files if I'm not wrong…

I just saw a big difference between static-tests and devtool usage, whereas there is just precommit as difference, the other big difference is the context used:
- static-tests is used in a context we do need only static-tests dependency
- devtool is used in a context we need to install tracim

This way what we are saying before about including requirements.txt in all requirements.txt doesn't make sens for static-tests specific case.

**Comment by inkhey** [comment]
> > Maybe should i change the mayway requirements.txt files are splitted :
> > requirements-static-tests is currently CI specific, what does exist in backend is testing and devtool, dev mean testing + devtool, maybe should i adapt static-tests to be same as devtool and create a requirements-tests.txt.
> 
> Yes I think this is a good idea, then dev will only be a shortcut (including) static-tests,tests and build. We'll get 7 requirements.txt files if I'm not wrong…

One other problem is see is that i'm not sure it make sense to include build in other requirements.txt file as if we do pip install pip with other package we are not sure the correct pip version will be used, so it make sense to have a least three step for python package install:
- build dep
- normal or dev dependencies
- package itself in editable mode

so i suggest that all but requirements-dev.txt which is a shortcut for : requirements.txt, requirements-test.txt, requirements-devtools.txt contain requirements.txt. 
This way, it's more clear and we can use devtools file for both ci and normal context.

**Comment by inkhey** [comment]
@grignards as you validated https://github.com/tracim/tracim/pull/3476,  you  can you (re)confirm that you're ok with the last changes (explain here : https://github.com/tracim/tracim/issues/3191#issuecomment-683820269 and here: https://github.com/tracim/tracim/issues/3191#issuecomment-683823312) : https://github.com/tracim/tracim/pull/3476/commits/76643e948742ff7a3e6699446c5a08839ac2efe6 , https://github.com/tracim/tracim/pull/3476/commits/a629bd97e99d0f29f2a68be5c44da4dcdae868c5 .

**Comment by PhilippeAccorsi** [comment]
@inkhey can you add tracim version where this issue is fixed ?
And add the issue in the correct project is still exist.

**Comment by inkhey** [comment]
this issue was fixed in 3.1.0.
